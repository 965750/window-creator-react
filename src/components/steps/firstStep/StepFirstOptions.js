import React, { Component } from 'react'
import { connect } from 'react-redux'
import {
    resizeWindow,
    changeWindowType,
} from '../../../store/actions/creatorActions'
import { setNotification } from '../../../store/actions/themeActions'
import styles from './StepFirstOptions.module.scss'
import BaseCheckbox from '../../blocks/BaseCheckbox'
import { FormattedMessage } from 'react-intl'

class StepFirstOptions extends Component {
    constructor(props) {
        super(props)
        this.heightInput = React.createRef()
        this.widthInput = React.createRef()
    }

    state = {
        doorType: 1,
    }

    componentDidMount() {
        this.heightInput.current.value = this.props.window.height
        this.widthInput.current.value = this.props.window.width
    }

    handleDoorType = value => {
        this.props.changeWindowType(value)
    }

    handleChange = e => {
        let value = Number(e.target.value)

        const limits = {
            height: {
                min: 160,
                max: 300,
            },
            width: {
                min: 60,
                max: 170,
            },
        }

        if (e.target.id === 'height') {
            value > limits.height.max
                ? (value = limits.height.max)
                : (value = value)
            value < limits.height.min
                ? (value = limits.height.min)
                : (value = value)
        } else if (e.target.id === 'width') {
            value > limits.width.max
                ? (value = limits.width.max)
                : (value = value)
            value < limits.width.min
                ? (value = limits.width.min)
                : (value = value)
        }

        if (
            this[`${e.target.id}Input`].current.value >
                limits[e.target.id].max ||
            this[`${e.target.id}Input`].current.value < limits[e.target.id].min
        ) {
            this.props.setNotification(
                `Window's height must be between ${limits.height.max}cm and ${limits.height.min}cm, also for width it's ${limits.width.max}cm and ${limits.width.min}cm`
            )

            this[`${e.target.id}Input`].current.value = value
        }

        this.props.resizeWindow({
            [e.target.id]: value,
            id: e.target.id,
        })
    }

    render() {
        return (
            <div className={`pt-2`}>
                <div className={`mb-10`}>
                    <p className="mb-4 text-darkText border-b border-mercury">
                        <FormattedMessage
                            id="Door type"
                            defaultMessage="Door type"
                        />
                    </p>
                    <div
                        className="mb-2 flex cursor-pointer"
                        onClick={() => this.handleDoorType(1)}
                    >
                        <BaseCheckbox
                            classes="mt-sm mr-2"
                            isChecked={this.props.window.doorType === 1}
                        />
                        <FormattedMessage
                            id="Single door"
                            defaultMessage="Single door"
                        />
                    </div>
                    <div
                        className="flex cursor-pointer"
                        onClick={() => this.handleDoorType(2)}
                    >
                        <BaseCheckbox
                            classes="mt-sm mr-2"
                            isChecked={this.props.window.doorType === 2}
                        />
                        <FormattedMessage
                            id="Double door"
                            defaultMessage="Double door"
                        />
                    </div>
                </div>
                <div>
                    <p className="mb-4 text-darkText border-b border-mercury">
                        <FormattedMessage
                            id="Door size"
                            defaultMessage="Door size"
                        />
                    </p>
                    <div className="flex mb-2">
                        <p className={`${styles.inputBox__label} mr-5`}>
                            <FormattedMessage
                                id="Width"
                                defaultMessage="Width"
                            />
                        </p>
                        <input
                            id="width"
                            className={`${styles.inputBox} border mr-1 text-center`}
                            onBlur={this.handleChange}
                            ref={this.widthInput}
                            type="number"
                        />
                        <span>cm</span>
                    </div>
                    <div className="flex">
                        <p className={`${styles.inputBox__label} mr-5`}>
                            <FormattedMessage
                                id="Height"
                                defaultMessage="Height"
                            />
                        </p>
                        <input
                            id="height"
                            className={`${styles.inputBox} border mr-1 text-center`}
                            onBlur={this.handleChange}
                            ref={this.heightInput}
                            type="number"
                        />
                        <span>cm</span>
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        window: state.creator.window,
    }
}

const mapDispatchToProps = dispatch => {
    return {
        resizeWindow: sizes => dispatch(resizeWindow(sizes)),
        changeWindowType: doorType => dispatch(changeWindowType(doorType)),
        setNotification: notification =>
            dispatch(setNotification(notification)),
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(StepFirstOptions)
