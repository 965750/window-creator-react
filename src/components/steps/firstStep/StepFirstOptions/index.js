import { connect } from 'react-redux'
import { FormattedMessage } from 'react-intl'
import {
  resizeWindow,
  changeWindowType,
} from '../../../../store/actions/creatorActions'
import { setNotification } from '../../../../store/actions/themeActions'

import BaseRadio from '../../../inputs/BaseRadio'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './StepFirstOptions.module.scss'

class StepFirstOptions extends Component {
  constructor(props) {
    super(props)
    this.heightInput = React.createRef()
    this.widthInput = React.createRef()
  }

  componentDidMount() {
    this.updateInputs()
  }

  componentDidUpdate(prevProps) {
    if (prevProps.window !== this.props.window) {
      this.updateInputs()
    }
  }

  updateInputs = () => {
    this.heightInput.current.value = this.props.window.height
    this.widthInput.current.value = this.props.window.width
  }

  handleDoorType = (value) => {
    this.props.changeWindowType(value)
  }

  handleChange = (e) => {
    let value = Number(e.target.value)
    const limits = {
      height: {
        min: 180,
        max: 260,
      },
      width: {
        min: 70,
        max: 140,
      },
    }

    if (e.target.id === 'height') {
      if (value > limits.height.max) value = limits.height.max
      if (value < limits.height.min) value = limits.height.min
    } else if (e.target.id === 'width') {
      if (value > limits.width.max) value = limits.width.max
      if (value < limits.width.min) value = limits.width.min
    }

    if (
      this[`${e.target.id}Input`].current.value > limits[e.target.id].max
      || this[`${e.target.id}Input`].current.value < limits[e.target.id].min
    ) {
      this.props.setNotification(
        `Window's height must be between ${limits.height.max}cm and ${limits.height.min}cm, also for width it's ${limits.width.max}cm and ${limits.width.min}cm`,
        'error',
      )

      this[`${e.target.id}Input`].current.value = value
    }

    this.props.resizeWindow({
      [e.target.id]: value,
      id: e.target.id,
    })
  }

  render() {
    return (
      <div className="pt-2">
        <div className="mb-10">
          <p className="mb-4 text-osloGray border-b border-mercury">
            <FormattedMessage id="Door type" defaultMessage="Door type" />
          </p>
          <button
            className="mb-2 flex cursor-pointer"
            onClick={() => this.handleDoorType(1)}
          >
            <BaseRadio
              classes="mt-sm mr-2"
              isChecked={this.props.window.doorType === 1}
            />
            <FormattedMessage id="Single door" defaultMessage="Single door" />
          </button>
          <button
            className="flex cursor-pointer"
            onClick={() => this.handleDoorType(2)}
          >
            <BaseRadio
              classes="mt-sm mr-2"
              isChecked={this.props.window.doorType === 2}
            />
            <FormattedMessage id="Double door" defaultMessage="Double door" />
          </button>
        </div>
        <div>
          <p className="mb-4 text-osloGray border-b border-mercury">
            <FormattedMessage id="Door size" defaultMessage="Door size" />
          </p>
          <div className="flex mb-2">
            <p className={`${styles.inputBox__label} mr-5`}>
              <FormattedMessage id="Width" defaultMessage="Width" />
            </p>
            <input
              data-testid="widthInput"
              id="width"
              className={`${styles.inputBox} border mr-1 text-center`}
              onBlur={this.handleChange}
              ref={this.widthInput}
              type="number"
            />
            <span>cm</span>
          </div>
          <div className="flex">
            <p className={`${styles.inputBox__label} mr-5`}>
              <FormattedMessage id="Height" defaultMessage="Height" />
            </p>
            <input
              data-testid="heightInput"
              id="height"
              className={`${styles.inputBox} border mr-1 text-center`}
              onBlur={this.handleChange}
              ref={this.heightInput}
              type="number"
            />
            <span>cm</span>
          </div>
        </div>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  resizeWindow: (sizes) => dispatch(resizeWindow(sizes)),
  changeWindowType: (doorType) => dispatch(changeWindowType(doorType)),
  setNotification: (notification, notificationType) => dispatch(setNotification(notification, notificationType)),
})

StepFirstOptions.propTypes = {
  window: PropTypes.shape({
    color: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    doorType: PropTypes.number,
    rows: PropTypes.array,
    columns: PropTypes.array,
  }),
  resizeWindow: PropTypes.func,
  changeWindowType: PropTypes.func,
  setNotification: PropTypes.func,
}

export default connect(
  null,
  mapDispatchToProps,
)(StepFirstOptions)
