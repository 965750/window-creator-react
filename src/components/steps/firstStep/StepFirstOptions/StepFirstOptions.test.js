import '@testing-library/jest-dom/extend-expect'

import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup, fireEvent } from '@testing-library/react'

import StepFirstOptions from './index'
import messages from '../../../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
  },
}

const reducer = (state = initialStore, action) => {
  switch (action.type) {
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('StepFirstOptions', () => {
  let mockFunc = jest.fn()
  let wrapper
  const initialProps = {
    window: {
      color: '#000',
      width: 110,
      height: 230,
      doorType: 2,
      rows: [0, 1, 2],
      columns: [0, 1, 2],
    },
    resizeWindow: jest.fn(),
    changeWindowType: jest.fn(),
    setNotification: jest.fn(),
  }

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <IntlProvider
        locale="en"
        messages={messages[store.getState().theme.lang]}
      >
        <StepFirstOptions {...props} />
      </IntlProvider>,
    )
  }

  afterEach(cleanup)

  it('should match snapshot', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should validate height input\'s maximum limit', () => {
    setUp()

    const input = wrapper.getByTestId('heightInput')

    fireEvent.change(input, {
      target: { value: 310 }
    })

    fireEvent.blur(input)
    expect(input.value).toBe("300")
  })

  it('should validate height input\'s minimum limit', () => {
    setUp()

    const input = wrapper.getByTestId('heightInput')

    fireEvent.change(input, {
      target: { value: 120 }
    })

    fireEvent.blur(input)
    expect(input.value).toBe("160")
  })

  it('should validate width input\'s maximum limit', () => {
    setUp()

    const input = wrapper.getByTestId('widthInput')

    fireEvent.change(input, {
      target: { value: 220 }
    })

    fireEvent.blur(input)
    expect(input.value).toBe("170")
  })

  it('should validate width input\'s minimum limit', () => {
    setUp()

    const input = wrapper.getByTestId('widthInput')

    fireEvent.change(input, {
      target: { value: 10 }
    })

    fireEvent.blur(input)
    expect(input.value).toBe("60")
  })
})
