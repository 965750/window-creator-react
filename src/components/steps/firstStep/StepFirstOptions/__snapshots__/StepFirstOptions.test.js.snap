// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepFirstOptions should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="pt-2"
  >
    <div
      class="mb-10"
    >
      <p
        class="mb-4 text-osloGray border-b border-mercury"
      >
        Typ drzwi
      </p>
      <button
        class="mb-2 flex cursor-pointer"
      >
        <div
          class="radio__wrapper border-botticelli border-4 rounded-full relative mt-sm mr-2"
          data-testid="radioWrapper"
        />
        Pojedyńcze drzwi
      </button>
      <button
        class="flex cursor-pointer"
      >
        <div
          class="radio__wrapper border-botticelli border-4 rounded-full relative mt-sm mr-2"
          data-testid="radioWrapper"
        >
          <div
            class="radio--checked top-half left-half rounded-full bg-hippieBlue absolute"
          />
        </div>
        Podwójne drzwi
      </button>
    </div>
    <div>
      <p
        class="mb-4 text-osloGray border-b border-mercury"
      >
        Wymiar drzwi
      </p>
      <div
        class="flex mb-2"
      >
        <p
          class="inputBox__label mr-5"
        >
          Szerokość
        </p>
        <input
          class="inputBox border mr-1 text-center"
          data-testid="widthInput"
          id="width"
          type="number"
        />
        <span>
          cm
        </span>
      </div>
      <div
        class="flex"
      >
        <p
          class="inputBox__label mr-5"
        >
          Wysokość
        </p>
        <input
          class="inputBox border mr-1 text-center"
          data-testid="heightInput"
          id="height"
          type="number"
        />
        <span>
          cm
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
